# Data Validation & Cleaning


### Importing pandas and load cvs file 

In [11]:
import pandas as pd 

games = pd.read_csv('mobile_game_inapp_purchases.csv')

games

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate
0,c9889ab0-9cfc-4a75-acd9-5eab1df0015c,49.0,Male,Norway,Android,Battle Royale,9,12.83,Minnow,11.40,28.0,Apple Pay,2025-03-19
1,7c9e413c-ecca-45f2-a780-2826a07952a2,15.0,Male,Switzerland,iOS,Action RPG,11,19.39,Minnow,6.37,18.0,Debit Card,2025-06-08
2,fd61e419-1a92-4f43-a8c7-135842ad328a,23.0,Male,China,Android,Fighting,9,8.87,Minnow,15.81,30.0,Apple Pay,2025-06-02
3,bdb7f6d1-ff9a-468c-afe7-43f32a94293e,31.0,Male,Mexico,Android,Racing,12,19.56,Minnow,13.49,9.0,Debit Card,2025-04-01
4,aa7eec14-4846-47b9-b879-9c98038cda04,37.0,Female,India,Android,Battle Royale,10,15.23,Minnow,10.86,15.0,Paypal,2025-05-05
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3019,29f9fea9-591d-4e9f-a662-958655d3eb4a,19.0,Female,Iran,Android,Fighting,7,34.81,Minnow,12.48,16.0,Google Pay,2025-05-28
3020,c906eb00-fb33-4bf2-97a4-311634cd0d7a,48.0,Male,Switzerland,iOS,Adventure,9,19.45,Minnow,14.77,14.0,Carrier Billing,2025-07-25
3021,26e17717-ebec-42a9-aae2-2eff66888119,44.0,Male,France,Android,Card,15,20.54,Minnow,16.69,24.0,Gift Card,2025-01-25
3022,4d6e2bfb-cb11-411a-942c-4fb7a4832088,45.0,Male,Sweden,iOS,Fighting,8,14.48,Minnow,17.29,15.0,Paypal,2025-05-07


### Counting NaNs

In [13]:
games.isna().sum()

UserID                             0
Age                               60
Gender                            60
Country                           60
Device                            60
GameGenre                         60
SessionCount                       0
AverageSessionLength               0
SpendingSegment                    0
InAppPurchaseAmount              136
FirstPurchaseDaysAfterInstall    136
PaymentMethod                    136
LastPurchaseDate                 136
dtype: int64

136 player records contain missing values across all purchase-related fields (InAppPurchaseAmount, PaymentMethod, FirstPurchaseDaysAfterInstall and LastPurchaseDate), while still being assigned a SpendingSegment.

Since the purchase information cannot be verified, these records will be excluded from analyses involving revenue and monetization metrics.

They were retained for gameplay and engagement analyses, as gameplay-related fields (SessionCount and AverageSessionLength) remain complete.

In [15]:
games[
(games['SpendingSegment'].isin(['Whale', 'Minnow', 'Dolphin'])) &
(games['InAppPurchaseAmount'].isna())]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate
11,a516f724-6b0f-438c-b20b-59f5f35ab06a,31.0,Male,Sweden,Android,Simulation,10,31.10,Dolphin,NaN,NaN,NaN,NaN
31,2868c7dc-a4bc-4143-8256-334e29ec7458,54.0,Female,USA,Android,Sandbox,13,21.86,Minnow,NaN,NaN,NaN,NaN
43,daa87ce2-9524-437b-96f5-d862e44ba2a6,20.0,Female,Egypt,iOS,Adventure,12,13.56,Minnow,NaN,NaN,NaN,NaN
50,9c3598c6-6cd4-4da2-81ed-a8053987aa80,33.0,Female,Japan,Android,Fighting,12,26.08,Minnow,NaN,NaN,NaN,NaN
78,d88da455-eacb-4fc0-b7d8-9ee22f1c0ffa,18.0,Male,Saudi Arabia,Android,Fighting,6,34.42,Minnow,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
2879,2c036233-9c0a-4a7f-b724-bbdcc5d91559,NaN,Female,Japan,iOS,Strategy,13,27.67,Minnow,NaN,NaN,NaN,NaN
2930,3048af6d-baae-4191-8bdf-817c23661235,49.0,Male,Mexico,Android,Battle Royale,7,29.93,Dolphin,NaN,NaN,NaN,NaN
2990,d33756cf-dc49-416e-bde2-603371708a21,23.0,Male,Japan,iOS,Role Playing,8,26.30,Minnow,NaN,NaN,NaN,NaN
3012,efd47f71-4363-483f-89ba-d87e277a3c2e,35.0,Male,Germany,iOS,Role Playing,11,23.41,Minnow,NaN,NaN,NaN,NaN


In [16]:
missing_purchase_record = games.loc[
    games['InAppPurchaseAmount'].isna(),
    'SpendingSegment'
].value_counts().reset_index().rename(columns={'count': 'MissingPurchaseRecords'})

missing_purchase_record

,SpendingSegment,MissingPurchaseRecords
0,Minnow,111
1,Dolphin,20
2,Whale,5


### Checking df's Shape

In [18]:
games.shape

(3024, 13)

### Getting df info in order to undertand datatypes

In [20]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3024 entries, 0 to 3023
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   UserID                         3024 non-null   object 
 1   Age                            2964 non-null   float64
 2   Gender                         2964 non-null   object 
 3   Country                        2964 non-null   object 
 4   Device                         2964 non-null   object 
 5   GameGenre                      2964 non-null   object 
 6   SessionCount                   3024 non-null   int64  
 7   AverageSessionLength           3024 non-null   float64
 8   SpendingSegment                3024 non-null   object 
 9   InAppPurchaseAmount            2888 non-null   float64
 10  FirstPurchaseDaysAfterInstall  2888 non-null   float64
 11  PaymentMethod                  2888 non-null   object 
 12  LastPurchaseDate               2888 non-null   o

### Missing values in purchase-related columns are expected and represent players who have never completed an in-app purchase.

### Checking uniqueness of UserID

In [23]:
games['UserID'].nunique()

3024

In [24]:
games[['Age','Gender','Country','Device','GameGenre']].isna().all(axis=1).sum()

0

In [25]:
games[['Age','Gender','Country','Device','GameGenre']].isna().any(axis=1).sum()

291

### Missing values in demographic variables were preserved. Since they represent less than 10% of the dataset and are unrelated to purchase behaviour, they were excluded automatically from analyses requiring those specific attributes rather than being removed or imputed.

# Duplicates Checks 

In [28]:
games['UserID'].duplicated().sum()

0

In [29]:
games.duplicated().sum()

0

### No duplicated player IDs or fully duplicated records were found. Each row represents a unique player, confirming the dataset granularity (one row = one player).

# Data Type conversion 

In [32]:
games 

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate
0,c9889ab0-9cfc-4a75-acd9-5eab1df0015c,49.0,Male,Norway,Android,Battle Royale,9,12.83,Minnow,11.40,28.0,Apple Pay,2025-03-19
1,7c9e413c-ecca-45f2-a780-2826a07952a2,15.0,Male,Switzerland,iOS,Action RPG,11,19.39,Minnow,6.37,18.0,Debit Card,2025-06-08
2,fd61e419-1a92-4f43-a8c7-135842ad328a,23.0,Male,China,Android,Fighting,9,8.87,Minnow,15.81,30.0,Apple Pay,2025-06-02
3,bdb7f6d1-ff9a-468c-afe7-43f32a94293e,31.0,Male,Mexico,Android,Racing,12,19.56,Minnow,13.49,9.0,Debit Card,2025-04-01
4,aa7eec14-4846-47b9-b879-9c98038cda04,37.0,Female,India,Android,Battle Royale,10,15.23,Minnow,10.86,15.0,Paypal,2025-05-05
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3019,29f9fea9-591d-4e9f-a662-958655d3eb4a,19.0,Female,Iran,Android,Fighting,7,34.81,Minnow,12.48,16.0,Google Pay,2025-05-28
3020,c906eb00-fb33-4bf2-97a4-311634cd0d7a,48.0,Male,Switzerland,iOS,Adventure,9,19.45,Minnow,14.77,14.0,Carrier Billing,2025-07-25
3021,26e17717-ebec-42a9-aae2-2eff66888119,44.0,Male,France,Android,Card,15,20.54,Minnow,16.69,24.0,Gift Card,2025-01-25
3022,4d6e2bfb-cb11-411a-942c-4fb7a4832088,45.0,Male,Sweden,iOS,Fighting,8,14.48,Minnow,17.29,15.0,Paypal,2025-05-07


In [33]:
games.dtypes

UserID                            object
Age                              float64
Gender                            object
Country                           object
Device                            object
GameGenre                         object
SessionCount                       int64
AverageSessionLength             float64
SpendingSegment                   object
InAppPurchaseAmount              float64
FirstPurchaseDaysAfterInstall    float64
PaymentMethod                     object
LastPurchaseDate                  object
dtype: object

In [34]:
df = games.copy() # a copy of our DataFrame has been created

In [35]:
df['LastPurchaseDate'] = pd.to_datetime(df['LastPurchaseDate'], format = '%Y-%m-%d') 
df['Age'] = df['Age'].astype('Int64')
df['FirstPurchaseDaysAfterInstall'] = df['FirstPurchaseDaysAfterInstall'].astype('Int64') # 'Int64' preserves the NaN
df['GameGenre'] = df['GameGenre'].astype('category')

In [36]:
df.dtypes

UserID                                   object
Age                                       Int64
Gender                                   object
Country                                  object
Device                                   object
GameGenre                              category
SessionCount                              int64
AverageSessionLength                    float64
SpendingSegment                          object
InAppPurchaseAmount                     float64
FirstPurchaseDaysAfterInstall             Int64
PaymentMethod                            object
LastPurchaseDate                 datetime64[ns]
dtype: object

In [37]:
df

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate
0,c9889ab0-9cfc-4a75-acd9-5eab1df0015c,49,Male,Norway,Android,Battle Royale,9,12.83,Minnow,11.40,28,Apple Pay,2025-03-19
1,7c9e413c-ecca-45f2-a780-2826a07952a2,15,Male,Switzerland,iOS,Action RPG,11,19.39,Minnow,6.37,18,Debit Card,2025-06-08
2,fd61e419-1a92-4f43-a8c7-135842ad328a,23,Male,China,Android,Fighting,9,8.87,Minnow,15.81,30,Apple Pay,2025-06-02
3,bdb7f6d1-ff9a-468c-afe7-43f32a94293e,31,Male,Mexico,Android,Racing,12,19.56,Minnow,13.49,9,Debit Card,2025-04-01
4,aa7eec14-4846-47b9-b879-9c98038cda04,37,Female,India,Android,Battle Royale,10,15.23,Minnow,10.86,15,Paypal,2025-05-05
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3019,29f9fea9-591d-4e9f-a662-958655d3eb4a,19,Female,Iran,Android,Fighting,7,34.81,Minnow,12.48,16,Google Pay,2025-05-28
3020,c906eb00-fb33-4bf2-97a4-311634cd0d7a,48,Male,Switzerland,iOS,Adventure,9,19.45,Minnow,14.77,14,Carrier Billing,2025-07-25
3021,26e17717-ebec-42a9-aae2-2eff66888119,44,Male,France,Android,Card,15,20.54,Minnow,16.69,24,Gift Card,2025-01-25
3022,4d6e2bfb-cb11-411a-942c-4fb7a4832088,45,Male,Sweden,iOS,Fighting,8,14.48,Minnow,17.29,15,Paypal,2025-05-07


In [38]:
games.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3024 entries, 0 to 3023
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   UserID                         3024 non-null   object 
 1   Age                            2964 non-null   float64
 2   Gender                         2964 non-null   object 
 3   Country                        2964 non-null   object 
 4   Device                         2964 non-null   object 
 5   GameGenre                      2964 non-null   object 
 6   SessionCount                   3024 non-null   int64  
 7   AverageSessionLength           3024 non-null   float64
 8   SpendingSegment                3024 non-null   object 
 9   InAppPurchaseAmount            2888 non-null   float64
 10  FirstPurchaseDaysAfterInstall  2888 non-null   float64
 11  PaymentMethod                  2888 non-null   object 
 12  LastPurchaseDate               2888 non-null   o

In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3024 entries, 0 to 3023
Data columns (total 13 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   UserID                         3024 non-null   object        
 1   Age                            2964 non-null   Int64         
 2   Gender                         2964 non-null   object        
 3   Country                        2964 non-null   object        
 4   Device                         2964 non-null   object        
 5   GameGenre                      2964 non-null   category      
 6   SessionCount                   3024 non-null   int64         
 7   AverageSessionLength           3024 non-null   float64       
 8   SpendingSegment                3024 non-null   object        
 9   InAppPurchaseAmount            2888 non-null   float64       
 10  FirstPurchaseDaysAfterInstall  2888 non-null   Int64         
 11  PaymentMethod    

### Converting appropriate columns to optimized data types reduced memory usage and improved overall efficiency.

# Data Consistency Checks

In [42]:
df['Age'].describe()

count       2964.0
mean     33.533738
std      11.992258
min           13.0
25%           23.0
50%           33.0
75%           44.0
max           54.0
Name: Age, dtype: Float64

In [43]:
df[(df['Age'] < 0) | (df['Age'] > 100)]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate


In [44]:
df['SessionCount'].describe()

count    3024.000000
mean       10.074735
std         3.115863
min         1.000000
25%         8.000000
50%        10.000000
75%        12.000000
max        22.000000
Name: SessionCount, dtype: float64

In [45]:
df[df['SessionCount'] < 0]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate


In [46]:
df['AverageSessionLength'].describe()

count    3024.000000
mean       20.073978
std         8.585208
min         5.010000
25%        12.680000
50%        20.315000
75%        27.420000
max        34.990000
Name: AverageSessionLength, dtype: float64

In [47]:
df[df['AverageSessionLength'] < 0]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate


In [48]:
df['InAppPurchaseAmount'].describe()

count    2888.000000
mean      102.582864
std       454.339708
min         0.000000
25%         5.987500
50%        11.975000
75%        17.762500
max      4964.450000
Name: InAppPurchaseAmount, dtype: float64

In [49]:
df[df['InAppPurchaseAmount'] < 0]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate


In [50]:
df['FirstPurchaseDaysAfterInstall'].describe()

count       2888.0
mean     15.384003
std       8.946191
min            0.0
25%            8.0
50%           16.0
75%           23.0
max           30.0
Name: FirstPurchaseDaysAfterInstall, dtype: Float64

In [51]:
df[df['FirstPurchaseDaysAfterInstall'] < 0]

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate


In [52]:
df['PaymentMethod'].value_counts(dropna=False)

PaymentMethod
Debit Card         433
Google Pay         431
Gift Card          422
Carrier Billing    417
Credit Card        410
Paypal             401
Apple Pay          374
NaN                136
Name: count, dtype: int64

In [53]:
df['SpendingSegment'].value_counts(dropna=False)

SpendingSegment
Minnow     2544
Dolphin     412
Whale        68
Name: count, dtype: int64

In [54]:
df['Gender'].value_counts(dropna=False)

Gender
Male      1810
Female    1098
NaN         60
Other       56
Name: count, dtype: int64

In [55]:
df['Device'].value_counts(dropna=False)

Device
Android    1738
iOS        1226
NaN          60
Name: count, dtype: int64

In [56]:
df['GameGenre'].value_counts(dropna=False)

GameGenre
Sandbox          219
Simulation       219
Sports           217
Casual           209
Puzzle           206
Card             202
MMORPG           198
Strategy         198
Role Playing     193
Action RPG       191
Racing           191
Battle Royale    189
Fighting         181
MOBA             181
Adventure        170
NaN               60
Name: count, dtype: int64

In [57]:
clean_df = df.copy() 

In [58]:
clean_df.to_csv('clean_mobile_game_dataset.csv', index=False) #Saving the cleaned csv

In [59]:
clean_df

,UserID,Age,Gender,Country,Device,GameGenre,SessionCount,AverageSessionLength,SpendingSegment,InAppPurchaseAmount,FirstPurchaseDaysAfterInstall,PaymentMethod,LastPurchaseDate
0,c9889ab0-9cfc-4a75-acd9-5eab1df0015c,49,Male,Norway,Android,Battle Royale,9,12.83,Minnow,11.40,28,Apple Pay,2025-03-19
1,7c9e413c-ecca-45f2-a780-2826a07952a2,15,Male,Switzerland,iOS,Action RPG,11,19.39,Minnow,6.37,18,Debit Card,2025-06-08
2,fd61e419-1a92-4f43-a8c7-135842ad328a,23,Male,China,Android,Fighting,9,8.87,Minnow,15.81,30,Apple Pay,2025-06-02
3,bdb7f6d1-ff9a-468c-afe7-43f32a94293e,31,Male,Mexico,Android,Racing,12,19.56,Minnow,13.49,9,Debit Card,2025-04-01
4,aa7eec14-4846-47b9-b879-9c98038cda04,37,Female,India,Android,Battle Royale,10,15.23,Minnow,10.86,15,Paypal,2025-05-05
...,...,...,...,...,...,...,...,...,...,...,...,...,...
3019,29f9fea9-591d-4e9f-a662-958655d3eb4a,19,Female,Iran,Android,Fighting,7,34.81,Minnow,12.48,16,Google Pay,2025-05-28
3020,c906eb00-fb33-4bf2-97a4-311634cd0d7a,48,Male,Switzerland,iOS,Adventure,9,19.45,Minnow,14.77,14,Carrier Billing,2025-07-25
3021,26e17717-ebec-42a9-aae2-2eff66888119,44,Male,France,Android,Card,15,20.54,Minnow,16.69,24,Gift Card,2025-01-25
3022,4d6e2bfb-cb11-411a-942c-4fb7a4832088,45,Male,Sweden,iOS,Fighting,8,14.48,Minnow,17.29,15,Paypal,2025-05-07


### After validating missing values, duplicate records, data types and consistency rules, no issues were found that could materially affect the analysis. The dataset is considered reliable for the subsequent exploratory analyses.